### PR TITLE
Consolidated ValueAccessor implementations

### PR DESCRIPTION
**Pull request by Idclip**
Signed-off-by: Nick Avramoussis <4256455+Idclip@users.noreply.github.com>

### CODE REVIEW

File: openvdb/openvdb/tree/ValueAccessor.h [hunk 1]
```diff
@@ -132,3134 +179,767 @@ class ValueAccessorBase
...
     int getValueDepth(const Coord& xyz) const
     {
-        LockT lock(mMutex);
-        return mCache.getValueDepth(xyz);
+        return this->evalFirstCached(xyz, [&](const auto node) -> int
+            {
+                using NodeType = std::remove_pointer_t<decltype(node)>;
+                assert(node);
+
+                if constexpr(std::is_same<RootNodeT, NodeType>::value) {
```
Comment (danrbailey) [inline openvdb/openvdb/tree/ValueAccessor.h:440]: There's a similar pattern to this used in the `isCached(const Coord& xyz)` method which fails on GCC9. Do you have any insight as to why this one succeeds and the other fails? Is it worth perhaps always pre-evaluating the conditional logic of these constexpr expressions given that there may be other compilers we're not testing against that fail in a similar way?

**Review by danrbailey — COMMENTED** [review]
A few minor comments / questions. This is a big improvement in eliminating all this duplicate code, but it is quite a major change. I hope this won't open the door to a bunch of relatively incoherent compiler errors or performance regressions on various compilers that don't correctly implement these new(ish) features of C++. :/

File: openvdb/openvdb/Types.h [hunk 2]
```diff
@@ -198,6 +198,45 @@ struct IsSpecializationOf<Template<Args...>, Template>: public std::true_type {}
 ////////////////////////////////////////
 
 
+/// @brief  Re-implementation of C++17's index_sequence and the helper alias
+///   make_index_sequence. This was introduced to fix an issue with clang's
+///   builtin implementation which treats template specializations of builtin
+///   templates differently when a subsequent parameter is dependent. The
+///   result is a resolution failure during partial specialization selection.
```
[the remark this answers is not in the document]
Comment (Idclip) [reply openvdb/openvdb/Types.h:205]: Yes and yes. I chose not to define branches for the GCC implementation though and just always use ours, it's a bit cleaner

File: openvdb/openvdb/tree/ValueAccessor.h [hunk 3]
```diff
@@ -132,3134 +179,767 @@ class ValueAccessorBase
...
+        }
     }
-    const LeafNodeT* probeLeaf(const Coord& xyz) const { return this->probeConstLeaf(xyz); }
-    //@}
 
-    /// Remove all nodes from this cache, then reinsert the root node.
-    void clear() override
+    /// @brief  Remove all the cached nodes and invalidate the corresponding
+    ///   hash-keys.
+    void clear() override final
```
[the remark this answers is not in the document]
Comment (Idclip) [reply openvdb/openvdb/tree/ValueAccessor.h:802]: As discussed in TSC, leaving this as is

File: openvdb/openvdb/tree/ValueAccessor.h (hunk 1 above)
Comment (Idclip) [reply openvdb/openvdb/tree/ValueAccessor.h:440]: I've had a look at this and don't think I need to change it. The stuff that is crashing takes the form:

```
                using NodeType = typename NodeLevelList::template Get<Idx>;
                // @warning  Putting this exp in the if statement crashes GCC9
                constexpr bool IsRoot = std::is_same<RootNodeT, NodeType>::value;
```

Basically a `TypeList` access followed by an `is_same`.

The stuff remaining takes the form:

```
                using NodeType = std::remove_pointer_t<decltype(node)>;
                assert(node);
                constexpr bool IsRoot = std::is_same<RootNodeT, NodeType>::value;
```

Where `node` is the lambda runtime argument. These are two very different instantiations so it makes sense why we'd see a crash in one and not the other. I will leave this as is for now.

**Comment by jmlait** [comment]
I deployed this on our build farm and all of our current tool compilers successfully compiled it....  So that's a pretty good sign.

**Pull request merged**